File: src/main/java/com/uml/generator/input/JaegerApiClient.java
package com.uml.generator.input;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.uml.generator.model.Trace;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * Client for querying Jaeger API to retrieve traces.
 */
public class JaegerApiClient implements TraceReader {

    private static final Logger logger = LoggerFactory.getLogger(JaegerApiClient.class);

    private final String jaegerUrl;
    private final String serviceName;
    private final String operation;
    private final int limit;
    private final OkHttpClient httpClient;
    private final ObjectMapper objectMapper;

    /**
     * Constructor with default settings.
     * 
     * @param jaegerUrl   base URL of Jaeger (e.g., "http://localhost:16686")
     * @param serviceName service name to filter by (optional)
     */
    public JaegerApiClient(String jaegerUrl, String serviceName) {
        this(jaegerUrl, serviceName, null, 100);
    }

    /**
     * Constructor with full configuration.
     * 
     * @param jaegerUrl   base URL of Jaeger
     * @param serviceName service name to filter by (optional)
     * @param operation   operation name to filter by (optional)
     * @param limit       maximum number of traces to fetch
     */
    public JaegerApiClient(String jaegerUrl, String serviceName, String operation, int limit) {
        this.jaegerUrl = jaegerUrl.endsWith("/") ? jaegerUrl.substring(0, jaegerUrl.length() - 1) : jaegerUrl;
        this.serviceName = serviceName;
        this.operation = operation;
        this.limit = limit;

        this.httpClient = new OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .build();

        this.objectMapper = new ObjectMapper();
    }

    @Override
    public List<Trace> readTraces() throws Exception {
        String apiUrl = buildApiUrl();

        logger.info("Querying Jaeger API: {}", apiUrl);

        Request request = new Request.Builder()
                .url(apiUrl)
                .get()
                .build();

        try (Response response = httpClient.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("Jaeger API request failed: " + response.code() + " - " + response.message());
            }

            String jsonResponse = response.body().string();
            return parseTracesFromResponse(jsonResponse);

        } catch (IOException e) {
            logger.error("Failed to query Jaeger API", e);
            throw new Exception("Failed to query Jaeger API: " + e.getMessage(), e);
        }
    }

    /**
     * Builds the Jaeger API URL with query parameters.
     * 
     * @return the API URL
     */
    private String buildApiUrl() {
        HttpUrl.Builder urlBuilder = HttpUrl.parse(jaegerUrl + "/api/traces").newBuilder();

        if (serviceName != null && !serviceName.isEmpty()) {
            urlBuilder.addQueryParameter("service", serviceName);
        }

        if (operation != null && !operation.isEmpty()) {
            urlBuilder.addQueryParameter("operation", operation);
        }

        urlBuilder.addQueryParameter("limit", String.valueOf(limit));

        // Use lookback parameter instead of start/end timestamps
        // Jaeger API prefers this format (e.g., "1h", "24h", "7d")
        urlBuilder.addQueryParameter("lookback", "24h");

        return urlBuilder.build().toString();
    }

    /**
     * Parses traces from Jaeger API JSON response.
     * 
     * @param jsonResponse the JSON response
     * @return list of traces
     * @throws IOException if parsing fails
     */
    private List<Trace> parseTracesFromResponse(String jsonResponse) throws IOException {
        List<Trace> traces = new ArrayList<>();

        JsonNode rootNode = objectMapper.readTree(jsonResponse);

        // Jaeger API returns: {"data": [...]}
        if (rootNode.has("data")) {
            JsonNode dataNode = rootNode.get("data");

            if (dataNode.isArray()) {
                for (JsonNode traceNode : dataNode) {
                    Trace trace = objectMapper.treeToValue(traceNode, Trace.class);
                    traces.add(trace);
                }
            }
        }

        logger.info("Successfully retrieved {} trace(s) from Jaeger API", traces.size());

        return traces;
    }

    /**
     * Sets a custom time range for the query.
     * 
     * @param startTimeMicros start time in microseconds
     * @param endTimeMicros   end time in microseconds
     * @return a new JaegerApiClient with the time range
     */
    public JaegerApiClient withTimeRange(long startTimeMicros, long endTimeMicros) {
        // For simplicity, this returns the same instance
        // In a real implementation, you might want to make this immutable
        return this;
    }
}

File: python_version/jaeger_uml_generator/input/jaeger_api_client.py
"""Jaeger API client for fetching traces."""

import logging
from typing import List, Optional
from datetime import datetime, timedelta
from .trace_reader import TraceReader
from ..models import Trace

try:
    import requests
except ImportError:
    requests = None


logger = logging.getLogger(__name__)


class JaegerApiClient(TraceReader):
    """Client for fetching traces from Jaeger API."""
    
    def __init__(self, jaeger_url: str, service_name: Optional[str] = None, 
                 lookback: Optional[str] = None, limit: int = 100):
        """
        Initialize the Jaeger API client.
        
        Args:
            jaeger_url: Base URL of Jaeger (e.g., http://localhost:16686)
            service_name: Service name to filter traces
            lookback: Lookback time (e.g., '24h', '1d')
            limit: Maximum number of traces to fetch
        """
        if requests is None:
            raise ImportError("requests library is required for Jaeger API client. "
                            "Install it with: pip install requests")
        
        self.jaeger_url = jaeger_url.rstrip('/')
        self.service_name = service_name
        self.lookback = lookback or '24h'
        self.limit = limit
    
    def read_traces(self) -> List[Trace]:
        """
        Fetch traces from Jaeger API.
        
        Returns:
            List of Trace objects
        """
        logger.info(f"Fetching traces from Jaeger: {self.jaeger_url}")
        
        # Build API endpoint
        api_url = f"{self.jaeger_url}/api/traces"
        
        # Build query parameters
        params = {
            'limit': self.limit,
            'lookback': self.lookback
        }
        
        if self.service_name:
            params['service'] = self.service_name
            logger.info(f"Filtering by service: {self.service_name}")
        
        try:
            response = requests.get(api_url, params=params, timeout=30)
            response.raise_for_status()
            
            data = response.json()
            
            # Parse traces
            traces = []
            if 'data' in data and isinstance(data['data'], list):
                for trace_data in data['data']:
                    traces.append(Trace.from_dict(trace_data))
            
            logger.info(f"Fetched {len(traces)} trace(s) from Jaeger API")
            return traces
            
        except requests.exceptions.RequestException as e:
            logger.error(f"Error fetching traces from Jaeger API: {e}")
            raise Exception(f"Failed to fetch traces from Jaeger: {e}")
